### ISSUE TITLE
Better support running subsets of tests

### ISSUE THREAD

**Issue by jrubinator** [houseabsolute/test-class-moose#91]
While using Test::Class::Moose, I realized that the arguments to Test::Class::Moose::Load are not just directories, but *roots of libraries* - they expect the underlying path structure to match package names. This prevents running only a subset of the test directory, because we run into namespace-collisions:

```
[12:57]$ cat t/runtests.t
use strict;
use warnings;

use Test::Class::Moose::Runner;
use Test::Class::Moose::Load 't/Test/Custom';

Test::Class::Moose::Runner->new->runtests;
[12:57]$ cat t/Test/Custom/Importer.pm
package Test::Custom::Importer;

use Test::Class::Moose;

sub test_ok :Test {
    unlike $INC{'Importer.pm'},
           qr{Test/Custom},
           "`require Importer` won't use this file";
}

1;
[12:57]$ prove t/runtests.t
t/runtests.t .. 1/1         #   Failed test '`require Importer` won't use this file'
        #   at t/Test/Custom/Importer.pm line 6.
        #                   't/Test/Custom/Importer.pm'
        #           matches '(?^:Test/Custom)'
    # Failed test 'test_ok'
...
```

How would we feel about enhancing Test::Class::Moose::Load to support something like the following:

```
use Test::Class::Moose::Load
    -libs => [ qw(t/tests) ],
    -dirs => [ qw(t/tests/this t/tests/that) ], # but not t/tests/the_other
;
```

I think the old use-line `use Test::Class::Moose::Load @libs` would be equivalent to the following (and we could continue to support it):

```
use Test::Class::Moose::Load -libs => \@libs, -dirs => \@libs;
# or to
use Test::Class::Moose::Load -libs => \@libs;
```

**Comment by autarch** [comment]
Couldn't we just pass `-dirs` _without_ `-libs`? That seems like a simpler option.

**Comment by jrubinator** [comment]
Maybe? I'm not sure exactly what you're suggesting. Something like the following?
```
use Test::Class::Moose::Load @libs_to_load -dirs => \@dirs_to_test
```

I don't *think* you're proposing to remove the lib-loading feature altogether :)

**Comment by jrubinator** [comment]
@autarch - I'm happy to (want to!) add this feature, but I want to make sure we're on the same page before I start. Any clarifications would be appreciated :)

**Comment by autarch** [comment]
Sorry, I misunderstood the issue. I took a closer look at TCM::Load and I think I understand what you're getting at now.

It's not clear to me from looking at TCM::Load how much of the implementation is incidental versus how much is really necessary. For example, is it really necessary to modify `@INC`? Could we at least just modify it with `local @INC = ...` instead?

... Some experimentation later ... yes, we need to modify `@INC` since when loading a class it may pass an `extends` param to TCM and we need to be able to find the parent. But modifying it with `local` seems to work.

I think this `local` trick could fix your problem. It could in theory break code that relied on TCM::Load modifying `@INC` but that's not a documented feature (and I really think it's a _misfeature_).

@Ovid - if you have any thoughts on this I'd love to hear them. My inclination is to rewrite this to just rewrite this to load the _files as files_ rather than trying to load them as packages. This just seems much simpler. The code is already doing `use $package ()` so I don't see that loading it via `require $file` would have a different effect.

I made a branch to demonstrate at https://github.com/houseabsolute/test-class-moose/tree/autarch/tcm-load-uses-require

The `load.t` test is failing but when I look closely at it I really have no idea what it's testing. I _think_ it might be testing that all the dirs under `t/loadlib` are added to `@INC` but why are we testing for that?

**Comment by jrubinator** [comment]
The `local` trick would fix my problem, but I think it _is_ sort of documented. The POD right now says TCM::Load replaces

```
#!/usr/bin/perl -T
use strict;
use warnings;

use Test::Class::Moose::Runner;

use lib 't/tests'; 
...
```

If we make `@INC` local, doesn't that mean we're not doing anything like the `use lib` statement anymore?

**Comment by jrubinator** [comment]
Any action plan if @Ovid doesn't chime in?

**Comment by autarch** [comment]
I nag him via email ;)

**Comment by Ovid** [comment]
Wouldn't a better plan be to make `_load` public and allow people to override it? It's been in place for years and people have been happy with it.

`_load` tries to get the package name, munges `@INC`, and then `eval "use $package ();`. If we break that down and let people override the bits they want, that seems cleaner than risking breaking the contract.

If we go that route, changing this to avoid the hardcoded `.pm` might be good:

```perl
sub is_test_class {
    my ( undef, $file, undef ) = @_;

    # By default, we only care about .pm files
    if ( $file =~ /\.pm$/ ) {
        return 1;
    }
    return;
}
```

I hardcoded the extension, but some people _still_ insist upon naming their test classes with a `.t` extension. This might give them the tools to fine-tune this behavior on their own?

**Comment by autarch** [comment]
@jrubinator - so thinking about this even more, why aren't you using Test::Class::Moose::CLI or the Role::CLI role to write your own runner? Those both trivially support testing just one or a few classes. From the docs for Runner:

```
# From the command line ...
$> prove -lv t/tcm.t :: --classes TestFor::Foo --classes TestFor::Bar --tags quick --tags db --jobs 8
```
You can also pass a filename to `--classes` like `t/testlib/TestFor/Foo.pm` and it does the right thing. I could easily extend it to also accept a directory and have it find all the `.pm` files in that tree.

**Comment by jrubinator** [comment]
@autarch - I think we just didn't know about the CLI modules. 😃 We have some custom options, but those should be handleable by subclassing. So if we began using the CLI stuff, then modifying it to support directory handling would make sense.

Though as a new user, it would surprise me greatly if TCM supported directory (rather than lib) loading through the CLI modules but not through Runner. So I'm not sure where this fits in with Ovid's preferences.

@Ovid - publicizing `load` for override would also work (particularly the `@INC`-munging). One other suggestion if we provide easier overrides but don't worry about fundamental behavior - referencing `@libraries` rather than `@directories` where appropriate could help with clarity and maintainability.

**Comment by Ovid** [comment]
> So I'm not sure where this fits in with Ovid's preferences.

My only preference here is finding the minimal changes necessary to support the needed behavior. Many people rely on this module and I like to be cautious :)

**Comment by jrubinator** [comment]
> I could easily extend it to also accept a directory and have it find all the .pm files in that tree.

I've been thinking about this a bit more, and it seems like this is actually the way to go. I don't see a good way to get TCM::Load to convert filenames into modules, at least not without porting over the `_test_lib_dirs` logic from TCM::Role::CLI.

Is that latter something that we should do? I agree that expanding Role::CLI to take directories solves our use case if we just add directories options there.

**Comment by autarch** [comment]
So this is done in #94, but tests are failing because of an issue with the latest Test::More. I expect that to be fixed soon, and then we can merge this and do a release. If you could confirm that this does what you need that'd be great.

**Comment by jrubinator** [comment]
Yes, I believe that would accomplish what we were hoping for. Thanks!!!

**Comment by autarch** [comment]
Fixed in v0.93.
